# 4.6 Combining & Exporting Data B

## Directions

### 1. Import the orders_products_combined dataframe from the pickle file you just saved.

### 2. Check the shape of the imported dataframe (it should be the same as the one you exported—always check!).

### 3. Determine a suitable way to combine the orders_products_combined dataframe with your products data set stored in your “Prepared Data” folder. 

### 4. Confirm the results of the merge using the merge flag.

### 5. Export the newly created dataframe as ords_prods_merge in a suitable format.

In [12]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 1. Import the orders_products_combined dataframe from the pickle file you just saved

In [15]:
path = r'C:\Users\IVON6\Instacart Basket Analysis'

In [17]:
path

'C:\\Users\\IVON6\\Instacart Basket Analysis'

In [21]:
# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

# Import dataset products_cleanData.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleanData.csv'), index_col = False)

### 2. Check the shape of the imported dataframe (it should be the same as the one you exported—always check!).

In [24]:
# Check the output
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [28]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [30]:
# Method 1: Delete Unnamed Columns
df_prods.drop('Unnamed: 0', axis=1, inplace=True)

In [32]:
# Check the output
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [36]:
df_ords_prods_combined.shape

(32434489, 11)

In [38]:
df_prods.shape

(49672, 5)

### 3. Determine a suitable way to combine the orders_products_combined dataframe with your products data set stored in your “Prepared Data” folder.

#### Merged both datasets with their share column "product_id".

In [42]:
df_ords_prods_merged = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

ValueError: Cannot use name of an existing column for indicator column

#### ValueError: Cannot use name of an existing column for indicator column: Because I did one merge, I now have a column called _merge in my dataset. That's why when I merge again, I cannot create yet another _merge.

#### Alternatively I can pass in a string such as indicator='exists' Then my new column that "indicates" how I joined will be called exists and I can select it by doing df['exists']

In [46]:
df_ords_prods_merged = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = 'exists')

In [51]:
df_ords_prods_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                     3.0        False                  5          0   both   
1                    20.0        False                  1          1   both   
2                     6.0        False                 20          0   both   
3                     NaN         True                 10          0   both   
4                     9.0        False                 11          1   both   

  exists  
0   both  
1   both  
2   both  
3   both  
4   both

In [55]:
df_ords_prods_merged.shape

(32404859, 16)

#### 4. Confirm the results of the merge using the merge flag.

In [58]:
df_ords_prods_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### The dataframe after merging has 32,404,859 rows, and each of those rows have information found in both input data sets.

#### 5. Export the newly created dataframe as ords_prods_merge in a suitable format.

In [63]:
# Export data to pkl
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))